File: src/tachikoma/projects/processor.py
"""Post-processor for committing and pushing project changes.

Spawns Haiku agents to commit changes in dirty submodules after each session.
"""

import asyncio
from pathlib import Path

from claude_agent_sdk import ClaudeAgentOptions, query
from loguru import logger

from tachikoma.post_processing import PostProcessor
from tachikoma.projects.git import is_dirty, list_submodules, push
from tachikoma.sessions.model import Session

_log = logger.bind(component="projects")

SUBMODULE_COMMIT_PROMPT = """You are a git commit agent.
Your task is to inspect a project repository and create cohesive, well-organized commits
for ALL changes.

## Instructions

1. First, run `git log --oneline -10` to learn this project's commit style and conventions.

2. Check for any commit instructions in the repo:
   - Look for CONTRIBUTING.md, CLAUDE.md, or similar files
   - Follow any project-specific commit guidelines

3. Run `git status` to see all uncommitted changes.

4. Run `git diff` to understand what changed.

5. Group the changes into cohesive sets by purpose/directory:
   - Group related changes together
   - Each group should represent one logical change

6. For each group, create a commit:
   - Use `git add <files>` to stage the files in that group
   - Use `git commit -m "<descriptive message>"` with a message that:
     - Describes what changed and why
     - Follows the project's own commit style (from step 1)
     - Is concise but informative

## Important Constraints

- ONLY use these git commands: `git status`, `git diff`, `git log`, `git add`, `git commit`
- Do NOT use: `git push`, `git branch`, `git checkout`, `git reset`, `git rebase`,
  `git merge`, `git stash`, or any other commands
- Never ask for confirmation — just make the commits
- Commit EVERYTHING that shows up in `git status`
- If there are no changes, do nothing

Remember: These commits will be pushed to the project's remote. Good commit
messages help other developers understand the project's history."""


class ProjectsProcessor(PostProcessor):
    """Post-processor for committing and pushing project submodule changes.

    Checks each registered submodule for uncommitted changes, spawns a Haiku
    agent per dirty submodule to create descriptive commits, then pushes to
    each submodule's remote. Runs in the pre_finalize phase before GitProcessor.
    """

    def __init__(self, cwd: Path, cli_path: str | None = None) -> None:
        """Initialize the processor.

        Args:
            cwd: The workspace directory.
            cli_path: Optional path to the Claude CLI binary.
        """
        self._cwd = cwd
        self._cli_path = cli_path

    async def process(self, session: Session) -> None:
        """Process dirty submodules: commit and push changes.

        Args:
            session: The closed session (not used, but required by interface).
        """
        # Discover submodules
        submodule_paths = await list_submodules(self._cwd)
        if not submodule_paths:
            _log.debug("No submodules found, skipping project processing")
            return

        # Check which ones are dirty (parallel)
        dirty_results = await asyncio.gather(
            *[self._is_dirty(path) for path in submodule_paths],
            return_exceptions=True,
        )

        # Filter to dirty submodules
        dirty_paths: list[str] = []
        for path, result in zip(submodule_paths, dirty_results, strict=True):
            if isinstance(result, Exception):
                _log.warning(
                    "Failed to check submodule status: path={path} err={err}",
                    path=path,
                    err=str(result),
                )
            elif result is True:
                dirty_paths.append(path)

        if not dirty_paths:
            _log.debug("No dirty submodules, skipping commit")
            return

        _log.info(
            "Processing dirty submodules: count={count} paths={paths}",
            count=len(dirty_paths),
            paths=dirty_paths,
        )

        # Commit and push each dirty submodule in parallel
        results = await asyncio.gather(
            *[self._commit_and_push(path) for path in dirty_paths],
            return_exceptions=True,
        )

        # Log failures
        for path, result in zip(dirty_paths, results, strict=True):
            if isinstance(result, Exception):
                _log.warning(
                    "Failed to process submodule: path={path} err={err}",
                    path=path,
                    err=str(result),
                )

    async def _is_dirty(self, path: str) -> bool:
        """Check if a submodule has uncommitted changes.

        Args:
            path: The submodule path.

        Returns:
            True if dirty, False if clean.
        """
        submodule_path = self._cwd / path
        return await is_dirty(submodule_path)

    async def _commit_and_push(self, path: str) -> None:
        """Commit changes in a submodule and push to remote.

        Args:
            path: The submodule path.
        """
        submodule_path = self._cwd / path

        # Spawn agent to create commits
        _log.debug("Spawning commit agent: path={path}", path=path)
        await self._query_and_consume(submodule_path)

        # Push to remote
        try:
            await push(submodule_path)
            _log.info("Pushed submodule changes: path={path}", path=path)
        except Exception as e:
            # Log but don't fail - changes are committed locally
            _log.warning(
                "Push failed, changes remain committed locally: path={path} err={err}",
                path=path,
                err=str(e),
            )

    async def _query_and_consume(self, cwd: Path) -> None:
        """Spawn a fresh agent and consume its response.

        Args:
            cwd: The working directory for the agent.
        """
        options = ClaudeAgentOptions(
            model="haiku",
            cwd=cwd,
            cli_path=self._cli_path,
            permission_mode="bypassPermissions",
        )

        # Fully consume the async iterator to ensure the agent completes
        async for _ in query(prompt=SUBMODULE_COMMIT_PROMPT, options=options):
            pass

File: src/extensions/projects/processor.ts
import { join } from "node:path";

import type { CommitAgent } from "../../git/commit-agent.ts";
import { PUSH_SUCCESS, type RebaseResolver } from "../../git/sync.ts";
import type { Logger } from "../../log.ts";
import type { ExchangeProcessor, GitApi, PostProcessor } from "../api.ts";
import { isAhead, isDirty, listSubmodules } from "./git.ts";

export interface ProjectsProcessorDeps {
  workspaceRoot: string;
  agent: CommitAgent;
  git: GitApi;
  resolver?: RebaseResolver;
}

export const projectFallbackMessage = (name: string, now = new Date()): string =>
  `Update ${name} files (${now.toISOString().slice(0, 10)})`;

/**
 * Push a project's commits to `origin` and log the outcome. Shared by the dirty
 * pass (after committing) and the clean-ahead pass (push only), so the
 * `smartPush` + result-classification tail lives in one place. A rebase conflict
 * during push is handed to the resolver (agent-driven) when one is supplied.
 */
const pushAndReport = async (
  git: GitApi,
  repoPath: string,
  path: string,
  log: Logger,
  successMessage: string,
  failureMessage: string,
  resolver?: RebaseResolver,
): Promise<void> => {
  const result = await git.smartPush(repoPath, "origin", "HEAD", { log, resolver });

  if (PUSH_SUCCESS.has(result)) {
    log.info({ path, result }, successMessage);
  } else {
    log.warn({ path, result }, failureMessage);
  }
};

const commitAndPush = async (
  workspaceRoot: string,
  git: GitApi,
  agent: CommitAgent,
  path: string,
  log: Logger,
  resolver?: RebaseResolver,
): Promise<void> => {
  const repoPath = join(workspaceRoot, path);
  const name = path.split("/").at(-1) ?? path;

  const subjects = await git.commitAll({
    agent,
    cwd: repoPath,
    fallbackMessage: projectFallbackMessage(name),
    log,
  });

  if (subjects.length > 0) log.info({ path, subjects }, "committed project changes");

  await pushAndReport(
    git,
    repoPath,
    path,
    log,
    "pushed project changes",
    "push failed — changes remain committed locally",
    resolver,
  );
};

/**
 * Push (no commit) a project already ahead of its remote — a background task or
 * earlier exchange may have committed without pushing.
 */
const pushProject = async (
  workspaceRoot: string,
  git: GitApi,
  path: string,
  log: Logger,
  resolver?: RebaseResolver,
): Promise<void> => {
  await pushAndReport(
    git,
    join(workspaceRoot, path),
    path,
    log,
    "pushed ahead project changes",
    "push failed — ahead commits remain local",
    resolver,
  );
};

/**
 * Pre-finalize post-processor: commits and pushes every dirty registered
 * project before the workspace commit runs (so submodule pointer updates land
 * in the same workspace commit pass), then pushes any clean project that still
 * sits ahead of its remote so committed changes never linger across sessions.
 * A rebase conflict during either push is handed to the resolver (agent-driven)
 * before falling back to abort.
 */
export const createProjectsProcessor = ({
  workspaceRoot,
  agent,
  git,
  resolver,
}: ProjectsProcessorDeps): PostProcessor => ({
  name: "projects-commit",
  phase: "preFinalize",

  async process({ log }) {
    const submodulePaths = await listSubmodules(workspaceRoot);

    if (submodulePaths.length === 0) {
      log.debug("no submodules found — skipping project processing");
      return;
    }

    const dirtyChecks = await Promise.allSettled(
      submodulePaths.map((path) => isDirty(join(workspaceRoot, path))),
    );

    const dirtyPaths: string[] = [];
    const cleanPaths: string[] = [];

    for (const [index, check] of dirtyChecks.entries()) {
      const path = submodulePaths[index] as string;

      if (check.status === "rejected") {
        log.warn({ path, err: check.reason }, "failed to check submodule status");
      } else if (check.value) {
        dirtyPaths.push(path);
      } else {
        cleanPaths.push(path);
      }
    }

    // Pass 1: commit + push dirty submodules. The commit also pushes any local
    // commits already ahead of the remote.
    if (dirtyPaths.length > 0) {
      log.info({ paths: dirtyPaths }, "processing dirty submodules");

      const results = await Promise.allSettled(
        dirtyPaths.map((path) => commitAndPush(workspaceRoot, git, agent, path, log, resolver)),
      );

      for (const [index, result] of results.entries()) {
        if (result.status === "rejected") {
          log.warn({ path: dirtyPaths[index], err: result.reason }, "failed to process submodule");
        }
      }
    }

    // Pass 2: a clean tree can still sit ahead of its remote — commits made
    // earlier (a background task, a prior exchange) that nothing has pushed.
    // Push those so project commits never linger across sessions.
    const aheadChecks = await Promise.allSettled(
      cleanPaths.map((path) => isAhead(join(workspaceRoot, path))),
    );

    const aheadPaths: string[] = [];

    for (const [index, check] of aheadChecks.entries()) {
      const path = cleanPaths[index] as string;

      if (check.status === "rejected") {
        log.warn({ path, err: check.reason }, "failed to check if submodule is ahead");
      } else if (check.value) {
        aheadPaths.push(path);
      }
    }

    if (aheadPaths.length === 0) {
      if (dirtyPaths.length === 0) {
        log.debug("no dirty or ahead submodules — skipping project processing");
      }
      return;
    }

    log.info({ paths: aheadPaths }, "pushing clean submodules ahead of their remote");

    const pushResults = await Promise.allSettled(
      aheadPaths.map((path) => pushProject(workspaceRoot, git, path, log, resolver)),
    );

    for (const [index, result] of pushResults.entries()) {
      if (result.status === "rejected") {
        log.warn({ path: aheadPaths[index], err: result.reason }, "failed to push ahead submodule");
      }
    }
  },
});

export interface ProjectsExchangeProcessorDeps {
  workspaceRoot: string;
  git: GitApi;
  log: Logger;
}

/**
 * Per-exchange safety commit for submodules: after each exchange, commit any
 * dirty registered project in one cheap deterministic commit — no agent, no
 * model call, no push. The agent-grouped commit and push stay at trunk close
 * (see `createProjectsProcessor`).
 */
export const createProjectsExchangeProcessor = ({
  workspaceRoot,
  git,
  log,
}: ProjectsExchangeProcessorDeps): ExchangeProcessor => ({
  name: "projects-exchange-commit",

  async process() {
    const submodulePaths = await listSubmodules(workspaceRoot);

    if (submodulePaths.length === 0) return;

    const dirtyChecks = await Promise.allSettled(
      submodulePaths.map((path) => isDirty(join(workspaceRoot, path))),
    );

    const dirtyPaths: string[] = [];

    for (const [index, check] of dirtyChecks.entries()) {
      const path = submodulePaths[index] as string;

      if (check.status === "rejected") {
        log.warn({ path, err: check.reason }, "failed to check submodule status");
      } else if (check.value) {
        dirtyPaths.push(path);
      }
    }

    if (dirtyPaths.length === 0) return;

    const results = await Promise.allSettled(
      dirtyPaths.map((path) =>
        git.commitAllDeterministic({
          cwd: join(workspaceRoot, path),
          message: projectFallbackMessage(path.split("/").at(-1) ?? path),
          log,
        }),
      ),
    );

    for (const [index, result] of results.entries()) {
      if (result.status === "rejected") {
        log.warn({ path: dirtyPaths[index], err: result.reason }, "failed to commit submodule");
      } else if (result.value.length > 0) {
        log.debug(
          { path: dirtyPaths[index], subjects: result.value },
          "per-exchange project commit",
        );
      }
    }
  },
});
